include/irr/asset/ICPUShader.h

```diff
@@ -2,185 +2,53 @@
     size_t conservativeSizeEstimate() const override 
     { 
-        if (m_glsl.size())
-            return m_glsl.size();
-        else if (m_spirvBytecode)
-            return m_spirvBytecode->conservativeSizeEstimate();
-        return 0u; //shouldnt ever reach this line
-    }
-    void convertToDummyObject() override { m_glsl.clear(); }
-
-    const ICPUBuffer* getSPIR_VBytecode() const { return m_spirvBytecode; };
-
-    void enableIntrospection();
-    const SIntrospectionData* getIntrospectionData(const SEntryPointStagePair& _entryPoint) const
-    {
-        auto found = m_introspectionCache.find(_entryPoint);
-        return found == m_introspectionCache.end() ? nullptr : &found->second;
-    }
-
-    const core::vector<SEntryPointStagePair>& getStageEntryPoints();
-
-    inline const IParsedShaderSource* getParsed() const
-    {
-        obtainSPIRV();
-        return m_parsed;
+        return m_code->getSize();
     }
+    void convertToDummyObject() override { }
 
-protected:
-    const core::vector<SEntryPointStagePair>& getStageEntryPoints(spirv_cross::Compiler& _comp);
-    void obtainSPIRV() const;
+    const ICPUBuffer* getSPVorGLSL() const { return m_code; };
+    bool containsGLSL() const { return m_containsGLSL; }
 
-protected:
-    IGLSLCompiler* m_glslCompiler;
-    //`mutable` implies that POINTER might be altered (specifically it can be null or not-null when object gets created)
-    //`const` (on the left side of `*`) implies that the OBJECT once created, is not modified
-    mutable const ICPUBuffer* m_spirvBytecode = nullptr;
-    mutable const IParsedShaderSource* m_parsed = nullptr;
-    std::string m_glsl;
-    std::string m_glslOriginFilename;
-
-    using CacheKey = SEntryPointStagePair;
-    core::unordered_map<CacheKey, SIntrospectionData> m_introspectionCache;
-    core::vector<SEntryPointStagePair> m_entryPoints;
+    //! Only relevant when containsGLSL returns true
+    const std::string& getGLSLEntryPoint() const { return m_glslEntryPoint; }
+    //! Only relevant when containsGLSL returns true
+    E_SHADER_STAGE getGLSLStage() const { return m_glslStage; }
 
 protected:
-    struct SIntrospectionPerformer
-    {
-        SIntrospectionData doIntrospection(spirv_cross::Compiler& _comp, const SEntryPointStagePair& _ep) const;
-        void shaderMemBlockIntrospection(spirv_cross::Compiler& _comp, impl::SShaderMemoryBlock& _res, uint32_t _blockBaseTypeID, uint32_t _varID, const core::unordered_map<uint32_t, const SIntrospectionData::SSpecConstant*>& _mapId2sconst) const;
-        size_t calcBytesizeforType(spirv_cross::Compiler& _comp, const spirv_cross::SPIRType& _type) const;
-
-        static void deinitIntrospectionData(SIntrospectionData& _data);
-        static void deinitShdrMemBlock(impl::SShaderMemoryBlock& _res);
-    };
+    //! Might be GLSL null-terminated string or SPIR-V bytecode (denoted by m_containsGLSL)
+    ICPUBuffer* m_code;
```

Comment:
`m_code` could store both SPIR-V and GLSL, using only `m_containsGLSL` to switch between them

Reply:
thats how it is ;)

Reply:
ok, so no need for an `std::stringc`, you can supply a utility function in `core::` to turn std::string&& into an `core::smart_refctd_ptr<core::ICPUBuffer>`

Bonus points if you use our ICPUBuffer derived class with allocator, so you could actually steal std::string's data + allocator